### ISSUE TITLE
Discussion: the road to 0.1

### ISSUE THREAD

**Issue by Gekkio** [Gekkio/imgui-rs#167]
I've been thinking about doing some big refactorings in the project and doing a big push towards a 0.1 release. There's a couple of main topics I'd like to mention here:

### Automatic generation of bindings

In `imgui-rs`, we don't currently use bindgen and all the `imgui-sys` bindings are maintained by hand, which sometimes leads to human errors (fixed one mistake today I made in the latest cimgui upgrade). bindgen has been suggested in the past (most recently in #139) but there has been some cons in the approach, which is why I haven't been very excited about the idea. However, `cimgui` recently switched to automatic generation, which greatly changes the dynamics around the C bindings. I've got a WIP branch testing an approach that is slightly different to the previously suggested approach:

* instead of binding directly to Dear ImGui which generates non-portable C++ bindings, bind to `cimgui` which gives us portable C bindings. There's some non-portable preprocessor ifdefs, but those don't seem to end up affecting the generated Rust bindings
* instead of generating bindings at build-time which adds LLVM as a hard dependency, just regenerate bindings manually after every `cimgui` upgrade. Since the generated bindings are portable, there's no need to generate them on the fly

Let's take a look at the cons I listed in #139:

* *breaks pretty much all existing code (because many imgui-sys types are used in safe code too)*. 
This still applies, because we're replacing the entire imgui-sys API with an automatically generated one
* *user code becomes a mix of idiomatic and non-idiomatic code (because imgui-sys is used in safe code too)*.
This can still apply, but my next suggestion about the roles of -sys and the main crate should help
* ~~using bindgen with build.rs makes compiling much more complicated (e.g. LLVM is required on Windows. Do the bindings work with both msvc and gnu ABIs?)~~
Things should keep working on Windows just like they do now
* ~~LLVM 3.9 requirement means imgui-rs won't work out-of-the-box on Debian stable or old Ubuntu 16.04 LTS~~ 
LLVM won't be a requirement to use `imgui-rs`
* ~~a lot of magic in build.rs~~. 
Some "magic configuration" is still needed, but we can depend on `cimgui` to handle most of these things
* ~~bindgen has some limitations with C++. Are we sure the bindings work on all platforms and continue to work without causing a maintenance nightmare?~~ 
C++ limitations don't apply. We'll depend on `cimgui` to provide correct C bindings and there's a possibility of errors there, but `cimgui` is used by other bindings too so there's less risk than if we maintained complicated configuration ourselves

As a summary, it looks like the main downsides are

1. Breakage of existing code. This is inevitable with 0.1 anyway
2. Complications caused by how imgui-sys and imgui are partitioned. This leads to my next topic ->

### Roles of imgui-sys and imgui

Right now, the main `imgui` crate uses many types from `imgui-sys` in its public API, and there's quite a lot of useful stuff, including documentation comments, in `imgui-sys`. This approach won't work with auto-generated bindings very well, because some things (like documentation comments) simply can't be added to the auto-generated stuff directly. This is why I'm experimenting with the idea of making `imgui-sys` to contain just the auto-generated bindings and nothing else. **This will make the API of imgui-sys very annoying to use**. However, we can then focus on providing a good API and documentation in the main crate. Some very basic "helpers" like extra ImVec2/ImVec4 conversions will still stay in `imgui-sys`, but all other manual code will be removed.

### Tightening safety holes

I intend to do a big "sweep" over the entire API and fix all the safety holes I find, but this will cause API changes. There's already some known holes and I'm sure there's even more. I can immediately think of two:

* UB because you can create multiple ImGuiContexts and break some invariants
* UB because you can have a normal reference to ImGuiStyle but accidentally mutate it and break basic Rust guarantees about references

### Adding rustdoc documentation

I intend to add Rustdoc documentation comments to most things in the `imgui` crate API. `imgui-sys` won't have any docs since the main point is to just export the auto-generated bindings.

**Comment by Gekkio** [comment]
One of the first things I've looked at is context handling. Dear ImGui seems to be heading to a direction where a context encapsulates all state, and all functions will at some point in the future require a context explicitly, not implicitly like right now. `imgui-rs` has been designed since the beginning around explicit contexts so looks like this design decision is still good for 0.1.

However, it's possible to cause undefined behaviour with current `imgui-rs`. All you need to do is:

```rust
let imgui1 = ImGui::init();
let imgui2 = ImGui::init();
// whoops, this allows UB, because we can do stuff with both contexts but in reality
// all functions modify just one context: the current context
```

In 0.1 there is a separation between an "active context" and a "suspended context". Only one context is active at a time, but you can have multiple suspended contexts and switch between them. We can guarantee this with a combination of runtime and compile-time checks.

So, the same thing is safe (although it panics) in 0.1:

```rust
let ctx1 = Context::new();
let ctx2 = Context::new(); // panic
```

If we really need more than one context, a better way is to start with suspended contexts, because we can safely create as many as we want, and activation returns `Result` instead of a panic:

```rust
let suspended1 = SuspendedContext::new();
let suspended2 = SuspendedContext::new(); // this is ok
let ctx1 = suspended1.activate().unwrap(); // yay, active context
// do stuff with ctx1
let suspended1 = ctx1.suspend(); // suspend the context
let ctx2 = suspended2.activate().unwrap(); // yay, active context
// do stuff with ctx2
```

All this adds a bit of extra complexity, but the common simple case is still as easy as before, and multiple contexts are supported if you really need them. 99% of `imgui-rs` users will just do:

```rust
let ctx = Context::new();
```

which is the same as the current

```rust
let imgui = ImGui::init();
```

**Comment by Gekkio** [comment]
The next thing is the style object: `ImGuiStyle`. At first it looks like a very straightforward struct of simple data, but there's some invariants that complicate matters a bit.

Comments in Dear ImGui code suggest that the `ImGuiStyle` must not be directly modified between `NewFrame` and `Render/EndFrame` calls, so basically the style is frozen during a frame. However, you are allowed to use the various push/pop style/color functions to change things. Internally these functions keep track of the original value and *mutate the ImGuiStyle directly*. This causes problems with `imgui-rs` because if we call one of those push/pop functions, we need to guarantee there are no references to `ImGuiStyle`.

In current `imgui-rs` version, this causes UB:

```rust
let style: &ImGuiStyle = ui.imgui().style();
let alpha: &f32 = &style.alpha;
ui.with_style_var(StyleVar::Alpha(0.5), || {
  // oops, the alpha field has been mutated (set to 0.5) while we're holding an immutable reference to it
});
```

Fixing this in 0.1 requires tightening some rules, but I don't know yet which approach to use. Here are some options I've thought about so far:

* make `Ui::with_style_var` take `&mut self` instead. This probably leads to big problems with API ergonomics
* disallow immutable references to the style completely. This works, but means that you can't access the style struct at all during a frame
* same as the previous idea, but allow reading style fields using getter functions that don't return references. This is safe, because the values are always copied and field references are not involved

**Comment by rasky** [comment]
@Gekkio what do you think about @nsf approach? Would you think it would be viable for 0.1?

**Comment by Gekkio** [comment]
@rasky, automatic generation of bindings has already been implemented for a long time in the 0.1-dev branch, so we already have a working approach

**Comment by Gekkio** [comment]
I've done a huge amount of work in the 0.1-dev branch, and will probably explain some changes here later. But right now, I'd appreciate if some people could test the 0.1-dev branch by running the `hello_gfx` and `hello_glium` examples in the renderer directories. The old examples don't work yet, but these two new ones should. Try also changing in the example code `HiDpiMode::Rounded` to `HiDpiMode::Locked(somevalue)` with some locked DPI values (e.g. 1.0, 2.0, also 2.0/3.0 to try fractional DPI). The UI should scale correctly, although fonts are still blurry because I haven't added support for font configuration yet.

Here's my own test results:

Desktop machine with discrete Nvidia:

:white_check_mark: Ubuntu X11: both gfx and glium work fine
:white_check_mark: Windows 10: both gfx and glium work fine

Laptop with integrated Intel + discrete Nvidia combo:

:white_check_mark: Ubuntu X11 (AFAIK I'm using Nvidia): both gfx and glium work fine
:cry: Ubuntu Wayland (AFAIK uses Intel): both gfx and glium are completely messed up with various issues. Unusable
:white_check_mark: Windows 10 (Nvidia forced from settings): both gfx and glium work fine
:white_check_mark: Windows 10 (Intel forced from settings): both gfx and glium work fine (had some issues but they were fixed with a driver update. Issues also happened with other OpenGL apps so it was clearly a driver problem)

**Comment by yanchith** [comment]
Is there any way to help with this besides testing the 0.1 dev branch?

**Comment by Gekkio** [comment]
> Is there any way to help with this besides testing the 0.1 dev branch?

Not really at this point, and with testing it's important to remember that things might not be fully functional yet, and things that are functional might still change. In any case, testing and any feedback about the current (0.0.x) or upcoming (0.1) API is very welcome.

I plan to do one or more beta releases once this work is near completion, so people can do more testing and give feedback.

**Comment by Gekkio** [comment]
Here's some important changes I've done for 0.1:

* `[f32; 2]` is used in all places where `ImVec2` would be used in the upstream C++ version. Same for `[f32; 4]` and `ImVec4`. This is a breaking change, because imgui-rs 0.0.x uses tuples (e.g. `(f32, f32)`) instead of arrays. The main reason for this change is that arrays have a well-defined memory layout, so it's (AFAIK) safe to directly transmute between `[f32; 2]` and `ImVec2`. Tuples are *not* safe for FFI use, which would break the next thing on this list
* Some structs (e.g. `Style` and `Io`) are now `repr(C)` structs that are directly transmuted from corresponding raw imgui-sys structs (with thorough memory layout unit testing of course). This means that there's no need for getters/setters in many cases and we can expose fields directly without extra work
* imgui-winit-support is no longer a collection of small utility functions and instead provides a more robust abstraction that should be much simpler to use (and easier to use correctly)
* initial (draft version) support for draw callbacks
* support for clipboard (just like upstream imgui, this requires some glue code for cross-platform support)
* imgui-gfx-renderer is parameterized over color format, so you can use it with `Rgba8` and `Srgba8`
* imgui-gfx-renderer does the gamma fix in the shader. Note: the colours are not completely correct, because sRGB handling in gfx is still messed up and blending ends up being incorrect (= anything with transparency will still have wrong colours)

I'm also trying to redesign some APIs to make them easier to use with Rust instead of just being dumb wrappers over the upstream APIs. For example, the radio button API in 0.0.x supports simple boolean radio buttons and radio buttons for `i32` values. But what about `u32` or other types? It turns out that wrapping the `i32` API is completely pointless and we can use generics to support easy radio buttons for any `Copy+PartialEq` types with just a tiny bit of Rust code. [Look at the relevant example](https://github.com/Gekkio/imgui-rs/blob/9e2d2fc50a2a8fd01a9a1326c4b6ab9e07da82b9/imgui-examples/examples/radio_button.rs#L63)

**Comment by Gekkio** [comment]
I've started hand-picking the changes from 0.1-dev to master, because maintaining a separate branch for a long time is way too complicated. I'll try to avoid removing any big functionality even if I didn't have that ready in 0.1-dev, so as a result we should have a working 0.1.0 with some breaking changes but it shouldn't have massive regressions in functionality.

In other words, (good) breaking changes are coming and 0.1.0 will be released soon but with less features than originally intended.

**Comment by Gekkio** [comment]
Closing because `0.1.0-rc.1` has been released and 0.1 is just behind the corner :wink: 
I didn't pull all the changes, so some things mentioned in this issue are not in 0.1 and might appear in 0.2 or later releases
